src/main/scala/components/csr.scala

```diff
@@ -0,0 +1,533 @@
+  val hpm19 = Bool()
+  val hpm18 = Bool()
+  val hpm17 = Bool()
+  val hpm16 = Bool()
+  val hpm15 = Bool()
+  val hpm14 = Bool()
+  val hpm13 = Bool()
+  val hpm12 = Bool()
+  val hpm11 = Bool()
+  val hpm10 = Bool()
+  val hpm9 = Bool()
+  val hpm8 = Bool()
+  val hpm7 = Bool()
+  val hpm6 = Bool()
+  val hpm5 = Bool()
+  val hpm4 = Bool()
+  val hpm3 = Bool()
+  val ir = Bool()
+  val tmzero = Bool()
+  val cy = Bool()
+}
+
+class MCause extends Bundle{
+  val interrupt = Bool()
+  val exceptioncode = UInt(31.W)
+}
+
+object MCSRCmd{
+  // commands
+  val size = 3.W
+  val execute = 0.asUInt(size)
+  val nop = 0.asUInt(size)
+  val write = 1.asUInt(size)
+  val set = 2.asUInt(size)
+  val clear = 3.asUInt(size)
+  val interrupt = 4.asUInt(size)
+  val read = 5.asUInt(size)
+  
+  val SIZE = 3.W
+  val MSB = 31
+  val LSB = 20
+  val TRAPADDR = "h80000000".U
+  val MPRV = 3
+}
+ 
+
+//reorder bundles
+class CSRRegFileIO extends Bundle{
+  //val hartid = Input(UInt(32.W))
+  val rw = new Bundle {
+    val rdata = Output(UInt(32.W)) //
+    val wdata = Input(UInt(32.W)) //
+  }
+
+  val csr_stall = Output(Bool())//not needed in single cycle
+  val eret = Output(Bool())// change ret names
+
+  val decode = new Bundle {
+    val inst = Input(UInt(32.W)) //
+    val immid = Input(UInt(32.W)) //
+    val read_illegal = Output(Bool())
+    val write_illegal = Output(Bool())
+    val system_illegal = Output(Bool())
+  }
+  
+  val regwrite = Output(Bool())
+  val status = Output(new MStatus())//not needed in this design but useful if more ISA extensions
+  val evec = Output(UInt(32.W)) //
+  val exception = Input(Bool())  // rename to illgl inst
+  val retire = Input(Bool()) // rename to retire inst
+  val pc = Input(UInt(32.W)) //
+  val time = Output(UInt(32.W))//
+}
+
+class CSRRegFile extends Module{
+  //INIT CSR
+  val io = IO(new CSRRegFileIO)
+  io := DontCare
+
+  val reset_mstatus = WireInit(0.U.asTypeOf(new MStatus()))
+  reset_mstatus.mpp := MCSRCmd.MPRV//machine mode
+  reset_mstatus.mie := true.B//machine mode
+  reset_mstatus.mie := true.B//machine mode
+
+  val reg_mstatus = RegInit(reset_mstatus)
+  val reg_mepc = Reg(UInt(32.W))
+  val reg_mcause = RegInit(0.U.asTypeOf(new MCause()))
+  val reg_mtval = Reg(UInt(32.W))
+  val reg_mscratch = Reg(UInt(32.W))
+  val reg_mtimecmp = Reg(UInt(64.W))
+  val reg_medeleg = Reg(UInt(32.W))
+
+  val reg_mip = RegInit(0.U.asTypeOf(new MIx()))
+  val reg_mie = RegInit(0.U.asTypeOf(new MIx()))
+  val reg_wfi = RegInit(false.B)
+  val reg_mtvec = RegInit(0.U.asTypeOf(new MTVec()))
+
+  val reg_time = WideCounter(64)
+  val reg_instret = WideCounter(64, io.retire)
+
+  val reg_mcounterinhibit = RegInit(0.U.asTypeOf(new XCounterEnInhibit()))
+  val reg_mcounteren = RegInit(0.U.asTypeOf(new XCounterEnInhibit()))
+
+  val read_mstatus = io.status.asUInt()
+  val isa_string = "I"
+  val reg_misa = RegInit((BigInt(0) | isa_string.map(x => 1 << (x - 'A')).reduce(_|_)).U.asTypeOf(new MISA()))
```

Comment:
Wow... the rocket chip people need to a little less crazy scala stuff. Could you at least write a comment on what this is doing? It might be better to just replace this with a hardcoded value (and a comment on what it means)

Reply:
fixed. hardcoded for I extension, but kept old code with comments just incase